content/blog/2019-12-18-novm/index.md

```diff
@@ -0,0 +1,317 @@
+The recursive sum benchmark ended up executing 6410250 different run-time allocations
+of 1544 bytes. While it had a very large performance impact, we could certainly tune
+the stack allocation algorithm to request larger chunks of memory to reduce the frequency
+of `malloc` system calls.
+
+### PARSEC Benchmarks
+
+While these microbenchmarks give us a good upper bound on worst-case overhead,
+we wanted to evaluate some more realistic tests. We chose the [PARSEC](https://parsec.cs.princeton.edu/)
+benchmarks, mostly because we used them for a prior project in this class and could test them easily.
+The execution times for these benchmarks are the built-in "Regions of Interest" and exclude initialization
+and warm up times for each program.
+
+<img src="stackcheckparsec.png" style="width:100%"/>
+
+With these four benchmarks, there was almost no impact of applying the split-stack option.
+We should note that the StreamCluster benchmark actually _sped up_ when using split-stack;
+likely this is some sort of memory alignment effect à la [Mytcowicz et al.](https://dl.acm.org/citation.cfm?id=1508275).
+In any case, we should probably consider this impact to be negligible.
+
+Of these benchmarks, only Ferret actually required dynamically allocating stack space.
+Each of these allocations was 618 KB, which is a potential concern. It is unclear, in a real
+system using only physical addressing, whether or not allocations of this size would
+be frequently servicable or not. I hypothesize that real systems with many gigabytes or terabytes
+of memory with even severe fragmentation will be able to regularly respond to allocations
+in the kilobyte range; however, proving this is future work.
+
+# Large Object Allocations
+
+The other major modification to programs would be supporting large memory allocations.
+Since it is probably unreliable to request very large contiguous memory regions, we must adopt a new
+strategy. To evaluate the potential impact of these changes, we modified the Blackscholes
+benchmark from the PARSEC suite. Blackscholes uses two dynamically allocated arrays,
+which we replaced with custom `Array` objects that we implemented to use only fixed size
+allocations. We chose to modify this application not only because it consists of a single,
+easily-modifiable source file, but also because it iterates through large arrays
+and is very likely to be negatively impacted by array access latency and spatial locality within the array.
+
+## Custom Array Implementation
+
+We implemented our `Array` using a tree-based datastructure that mimics
+the functionality of page tables in a virtual address environment.
+We support up to three levels of allocation, where the final
+level contains data and all previous levels contain pointers to
+other pages. In the following diagram, the **L1** page contains
+pointers to **L2** pages, which contain pointers to the **L3** pages.
+Each **L3** page contains the actual object array data.
```

Comment:
You might be interested to read about ["z-rays"](http://www.cs.utexas.edu/~mckinley/papers/zrays-pldi-2010.pdf).